[package] doxygen/1.9.4 doesn't build with gcc 12 or clang 15

### Description

Looks like doxygen 1.9.4 won't build with gcc12/clang15 as per https://github.com/doxygen/doxygen/issues/9312. The error thrown is:

```
[ 49%] Building CXX object addon/doxywizard/CMakeFiles/doxywizard.dir/moc_inputstrlist.cpp.o
cd /home/iurt/rpmbuild/BUILD/doxygen-1.9.4/build/addon/doxywizard && /usr/bin/c++ -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NO_CAST_FROM_ASCII -DQT_NO_CAST_TO_ASCII -DQT_NO_DEBUG -DQT_WIDGETS_LIB -DQT_XML_LIB -DUNICODE -I/home/iurt/rpmbuild/BUILD/doxygen-1.9.4/addon/doxywizard/. -I/home/iurt/rpmbuild/BUILD/doxygen-1.9.4/libversion -I/home/iurt/rpmbuild/BUILD/doxygen-1.9.4/build/generated_src -isystem /usr/include/qt5 -isystem /usr/include/qt5/QtCore -isystem /usr/lib64/qt5/mkspecs/linux-g++ -isystem /usr/include/qt5/QtGui -isystem /usr/include/qt5/QtWidgets -isystem /usr/include/qt5/QtXml -O2 -g -pipe -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector --param=ssp-buffer-size=4 -fasynchronous-unwind-tables -DNDEBUG -fvisibility=hidden -fvisibility-inlines-hidden -fPIC -std=gnu++17 -MD -MT addon/doxywizard/CMakeFiles/doxywizard.dir/moc_inputstrlist.cpp.o -MF CMakeFiles/doxywizard.dir/moc_inputstrlist.cpp.o.d -o CMakeFiles/doxywizard.dir/moc_inputstrlist.cpp.o -c /home/iurt/rpmbuild/BUILD/doxygen-1.9.4/build/addon/doxywizard/moc_inputstrlist.cpp
In file included from /home/iurt/rpmbuild/BUILD/doxygen-1.9.4/src/doxygen.h:24,
                 from /home/iurt/rpmbuild/BUILD/doxygen-1.9.4/src/message.cpp:21:
/home/iurt/rpmbuild/BUILD/doxygen-1.9.4/src/cache.h: In member function 'V* Cache<K, V>::insert(const K&, V&&)':
/home/iurt/rpmbuild/BUILD/doxygen-1.9.4/src/cache.h:53:14: error: 'exchange' is not a member of 'std'
   53 |         std::exchange(it->second->second,value);
      |              ^~~~~~~~
```

This isn't a problem with the recipe per se, and this is fixed in 1.9.5 so https://github.com/conan-io/conan-center-index/pull/16310 will provide a working solution.

The proposal for this ticket is to backport the patch since it's relatively trivial: https://github.com/doxygen/doxygen/commit/5198966c8d5fec89116d025c74934ac03ea511fa, which would give us a working doxygen 1.9.4 with more modern compilers.

### Package and Environment Details

NA

### Conan profile

```
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=20
compiler.libcxx=libstdc++11
compiler.version=12
os=Linux
[conf]
tools.build:compiler_executables={'c': '/usr/bin/gcc', 'cpp': '/usr/bin/g++'}
```

### Steps to reproduce

```
$ conan install --requires=doxygen/1.9.4 --build=missing
```

### Logs

<details><summary>Click to expand log</summary>

```
NA
```

</details>

--- a/recipes/doxygen/all/conandata.yml
+++ b/recipes/doxygen/all/conandata.yml
@@ -22,6 +22,7 @@ patches:
     - patch_file: "patches/1.9.4-0001-enable-modern-compilers.patch"
       patch_description: "Enable modern compilers"
       patch_type: "portability"
+      patch_source: "https://github.com/doxygen/doxygen/commit/5198966c8d5fec89116d025c74934ac03ea511fa"
   "1.9.2":
     - patch_file: "patches/1.9.2-0001-imported-target.patch"
       patch_description: "Fix includes"

--- a/recipes/doxygen/all/conandata.yml
+++ b/recipes/doxygen/all/conandata.yml
@@ -18,6 +18,10 @@ sources:
     url: "https://sourceforge.net/projects/doxygen/files/doxygen-1.8.17.src.tar.gz"
     sha256: "2cba988af2d495541cbbe5541b3bee0ee11144dcb23a81eada19f5501fd8b599"
 patches:
+  "1.9.4":
+    - patch_file: "patches/1.9.4-0001-enable-modern-compilers.patch"
+      patch_description: "Enable modern compilers"
+      patch_type: "bugfix"
   "1.9.2":
     - patch_file: "patches/1.9.2-0001-imported-target.patch"
       patch_description: "Fix includes"

--- a/recipes/doxygen/all/conandata.yml
+++ b/recipes/doxygen/all/conandata.yml
@@ -21,7 +21,7 @@ patches:
   "1.9.4":
     - patch_file: "patches/1.9.4-0001-enable-modern-compilers.patch"
       patch_description: "Enable modern compilers"
-      patch_type: "bugfix"
+      patch_type: "portability"
   "1.9.2":
     - patch_file: "patches/1.9.2-0001-imported-target.patch"
       patch_description: "Fix includes"

--- a/recipes/doxygen/all/patches/1.9.4-0001-enable-modern-compilers.patch
+++ b/recipes/doxygen/all/patches/1.9.4-0001-enable-modern-compilers.patch
@@ -1,0 +1,25 @@
+From 5198966c8d5fec89116d025c74934ac03ea511fa Mon Sep 17 00:00:00 2001
+From: Dimitri van Heesch <doxygen@gmail.com>
+Date: Fri, 6 May 2022 09:55:16 +0200
+Subject: [PATCH] issue #9312: Build: cache.h:53:14: error: 'exchange' is not a
+ member of 'std'
+
+---
+ src/cache.h | 1 +
+ 1 file changed, 1 insertion(+)
+
+diff --git a/src/cache.h b/src/cache.h
+index 0ff3092ed5..e218eb27d2 100644
+--- a/src/cache.h
++++ b/src/cache.h
+@@ -19,6 +19,7 @@
+ #include <list>
+ #include <unordered_map>
+ #include <mutex>
++#include <utility>
+ #include <ctype.h>
+ 
+ /*! Fixed size cache for value type V using keys of type K.
+-- 
+2.39.1
+

